### PR TITLE
Cardiac Reorientation Functions

### PR DESCRIPTION
**Pull request by MaziarSabouri**
cardiac_spect.py has been uploaded and spatial has been updated.

### CODE REVIEW

File: src/pytomography/utils/cardiac_spect.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.ndimage import affine_transform
+from scipy.ndimage import center_of_mass, shift
+import torch
+
```
Comment (lukepolson) [inline src/pytomography/utils/cardiac_spect.py:6]: remove line

File: src/pytomography/utils/cardiac_spect.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    slopes = np.tan(angles_rad)
+    intercepts = center_of_image[1] - slopes * center_of_image[0]
+    total_distances = np.sum(np.abs(slopes[:, None] * x_indices - y_indices + intercepts[:, None]) / np.sqrt(slopes[:, None]**2 + 1), axis=1)
+    best_idx = np.argmin(total_distances)
+    best_angle, best_slope, best_intercept = angles_rad[best_idx], slopes[best_idx], intercepts[best_idx]
+    x_line = np.linspace(0, shifted_binary_mask.shape[1], 500)
+    y_line = best_slope * x_line + best_intercept
+    angle_display = 90 + np.rad2deg(best_angle)
+
+    plt.imshow(shifted_binary_mask, cmap='gray')
```
Comment (lukepolson) [inline src/pytomography/utils/cardiac_spect.py:29]: Remove any plotting code `plt` from function

File: src/pytomography/utils/cardiac_spect.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    y_line = best_slope * x_line + best_intercept
+    angle_display = -1 * np.rad2deg(best_angle)
+    plt.imshow(mirrored_mask_horizontal, cmap='gray')
+    plt.plot(x_line, y_line, color='red', linestyle='--', linewidth=2)
+    plt.scatter(center_of_image[1], center_of_image[0], color='yellow', marker='x', label='Center of Mass')
+    plt.title(f"Shifted Binary Mask with Fitted Line at {angle_display:.1f}°")
+    plt.legend()
+    plt.axis('equal')
+    plt.tight_layout()
+    plt.show()
```
Comment (lukepolson) [inline src/pytomography/utils/cardiac_spect.py:75]: Remove

File: src/pytomography/utils/cardiac_spect.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+    tz = shift_values[1]
+    beta = angle_display
+    
+    result_volume = object_reorient(volume, tx, ty, tz, beta, alpha, gamma)
+    plot_slices(result_volume, 'Reoriented Volume')
+    
+    print(f"Translation (tx, ty, tz): ({tx}, {ty}, {tz})")
+    print(f"Rotation (beta, alpha, gamma): ({beta}, {alpha}, {gamma})")
```
Comment (lukepolson) [inline src/pytomography/utils/cardiac_spect.py:84]: remove print statements

File: src/pytomography/utils/cardiac_spect.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    angle_display = 90 + np.rad2deg(best_angle)
+
+    plt.imshow(shifted_binary_mask, cmap='gray')
+    plt.plot(x_line, y_line, color='red', linestyle='--', linewidth=2)
+    plt.scatter(center_of_image[1], center_of_image[0], color='yellow', marker='x', label='Center of Mass')
+    plt.title(f"Shifted Binary Mask with Fitted Line at {angle_display:.1f}°")
+    plt.legend()
+    plt.axis('equal')
+    plt.tight_layout()
+    plt.show()
```
Comment (lukepolson) [inline src/pytomography/utils/cardiac_spect.py:36]: remove

File: src/pytomography/utils/spatial.py [hunk 6]
```diff
@@ -141,4 +141,78 @@ def unpad_object_z(object: torch.Tensor, pad_size: int):
...
+    
+    for ax, slice_img, label in zip(axes, slices, ['HLA', 'SA', 'VLA']):
+        print(f"{label} slice shape: {slice_img.shape}")  # Debugging print
+        ax.imshow(slice_img, cmap='jet')
+        ax.set_title(label)
+        ax.axis('on')
+    
+    plt.suptitle(title)
+    plt.tight_layout()
+    plt.show()
```
Comment (lukepolson) [inline src/pytomography/utils/spatial.py:167]: move to notebook

File: src/pytomography/utils/spatial.py [hunk 7]
```diff
@@ -141,4 +141,78 @@ def unpad_object_z(object: torch.Tensor, pad_size: int):
...
+
+
+
+def object_reorient(volume, tx, ty, tz, beta, alpha, gamma):
```
Comment (lukepolson) [inline src/pytomography/utils/spatial.py:171]: Add object types to input arguments. Rename to `euler_angle_transform`

File: src/pytomography/utils/spatial.py [hunk 8]
```diff
@@ -141,4 +141,78 @@ def unpad_object_z(object: torch.Tensor, pad_size: int):
...
+    transform_matrix[:3, 3] = translation
+    
+    affine_matrix = transform_matrix[:3, :4]
+    
+    reoriented_volume = affine_transform(volume, affine_matrix, offset=translation, order=3)
+    
+    # Add a batch dimension
+    reoriented_volume = np.expand_dims(reoriented_volume, axis=0)
+    
+    return reoriented_volume
```
Comment (lukepolson) [inline src/pytomography/utils/spatial.py:218]: convert to `torch.tensor` and put on `pytomography.device`

`torch.tensor(reoriented_volume).to(pytomography.dtype).to(pytomography.device)`

File: src/pytomography/utils/spatial.py (hunk 7 above)
Comment (lukepolson) [reply src/pytomography/utils/spatial.py:171]: All input arguments are pytorch tensors

Comment (lukepolson) [reply src/pytomography/utils/spatial.py:171]: Rename volume to object

File: src/pytomography/utils/cardiac_spect.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.ndimage import affine_transform
+from scipy.ndimage import center_of_mass, shift
+import torch
+  
+  
+def reorientParams(volume, lowerThreshold=0.55, upperThreshold=1):
```
Comment (lukepolson) [inline src/pytomography/utils/cardiac_spect.py:8]: rename volume to object and expect pytorch.Tensor

**Review by lukepolson — CHANGES_REQUESTED** [review]

File: src/pytomography/utils/spatial.py (hunk 8 above)
Comment (MaziarSabouri) [reply src/pytomography/utils/spatial.py]: Done.

File: src/pytomography/utils/spatial.py (hunk 7 above)
Comment (MaziarSabouri) [reply src/pytomography/utils/spatial.py]: Done.

File: src/pytomography/utils/spatial.py (hunk 6 above)
Comment (MaziarSabouri) [reply src/pytomography/utils/spatial.py]: Removed from spatial.py file.

**Review by lukepolson — APPROVED** [review]

**Pull request merged**